File: ccEncryptionAssist.c
//Coordinate Code Encryption Assist v2.2.5 by TheMan132435421

//libraries
#include <stdio.h>
#include <cs50.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
//#include <daniilkvyat.h>

//strlen but for numbers
unsigned short numberstrlen(int input);
unsigned short numberstrlen(int input)
{
    unsigned short output=0;

    if(input < 0)
    {
        output++;
    }

    do
    {
        input=input / 10;
        output++;
    }
    while(input <= -1 || input >= 1);

    return output;
}

//removes punctuation from a string of text
string removePunctuation(string text);
string removePunctuation(string text)
{
    for(int t=0; t<strlen(text); t++)
    {
        //if current character is alphanumeric, lowercase it
        if(isalnum(text[t]))
        {
            text[t]=tolower(text[t]);
        }
        else if(text[t]==39 || text[t]==96) //if apostrophe, replace aprostrophes with something that no one uses
        {
            text[t]=94;
        }
        else //if not apostrophe, turn character into a space
        {
            text[t]=' ';
        }
    }

    //output formatted text
    return text;
}


//SECURITY (as a function)
short login(string enteredPassword);
short login(string enteredPassword)
{
    string passwords[]={"1", "12345", "password", "jkldfasjklfdajklfejkhfejhkfeqwjkhqeqwhjkfehgjferhjgerhjgfergferjhgfehjgfajhgfeawrhjgfeajhgefhjk"};

    short passwords_length=sizeof passwords / sizeof passwords[0];


    bool passkeys[passwords_length];
    for(short i=0; i<passwords_length; i++)
    {
        passkeys[i]=true;
    }

    for(short p=0; p<passwords_length; p++)
    {
        for(short i=0; i<(short)strlen(passwords[p]); i++)
        {
            if(enteredPassword[i]!=passwords[p][i])
            {
                passkeys[p]=false;
            }
        }
        if(strlen(enteredPassword)!=strlen(passwords[p]))
        {
            passkeys[p]=false;
        }
    }



    short fpc=passwords_length-1;

    for(short i=0; i<passwords_length-1; i++)
    {
        if(passkeys[i]==false)
        {
            fpc--;
        }
    }


    if(fpc<=0 && passkeys[passwords_length-1]==false)
    {
        return 1;
    }

    if(passkeys[passwords_length-1]==false)
    {
        return -1;
    }

    return 0;
}

//requires numberstrlen
int writeData(FILE* file, int retCode, int argc, string argv[]);
int writeData(FILE* file, int retCode, int argc, string argv[])
{
    //time info
    time_t orig_format;
    time(&orig_format);

    fprintf(file, "%i", retCode);
    for(short i=0; i<5-numberstrlen(retCode); i++)
    {
        fprintf(file, " ");
    }

    if(argc <= 999)
    {
        fprintf(file, "%i", argc);
        for(short i=0; i<5-numberstrlen(argc); i++)
        {
            fprintf(file, " ");
        }
    }
    else
    {
        fprintf(file, "999  ");
    }

    if(strlen(argv[0])<=30)
    {
        fprintf(file, "%s", argv[0]);
        for(short i=0; i<32-strlen(argv[0]); i++)
        {
            fprintf(file, " ");
        }
    }
    else
    {
        for(short i=0; i<30; i++)
        {
            fprintf(file, "%c", argv[0][i]);
        }
        fprintf(file, "  ");
    }

    fprintf(file, "%s", asctime(gmtime(&orig_format)));

    return 0;
}




////////////////////////////////////////////////////////////////////

int main(int argc, string argv[])
{
    //timekeeping from https://fresh2refresh.com/c-programming/c-time-related-functions/
    time_t orig_format;
    time(&orig_format);

    //pointer to data file
    FILE *info;
    string f=".data";

    //reject invalid amount of command-line arguments
    if(argc>2)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 1, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }

        printf("Error: expected './ccEncryptionAssist' or './ccEncryptionAssist n'\n");
        return 1;
    }

    //user-entered password
    string password=get_string("Enter password: ");
    //set user login status using security function
    short loginStatus=login(password);


    //if wrong password, exit program
    if(loginStatus>0)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 2, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }

        printf("Incorrect password.\n");
        return 2;
    }


    //if the program was not stopped, continue
    printf("Access granted. ");
    //announce any special conditions
    if(argc>1)
    {
        printf("Printing each ordered pair on its own line.\n\n");
    }
    else
    {
        printf("Note: enter './ccEncryptionAssist n' to print each ordered pair on a new line\n\n");
    }

    /////////////////////////////////////////////////////////////////////////////////////////////////////////

    //INPUT

    //ask user for key and input
    string key=get_string("Enter key: ");
    printf("\n");

    string plaintext=get_string("Enter plaintext: ");


    //Bug fix: loop through the key and plaintext, replacing any characters that can interfere with the word count and making everything lowercase
    key=removePunctuation(key);



    //check for buffer overruns
    if(loginStatus<0)
    {
        info=fopen(f, "a");
        if(info!=NULL)
        {
            writeData(info, 83, argc, argv);
            fclose(info);
        }
        else
        {
            printf("**FATAL ERROR\n");
            return 1;
        }

        for(short i=0; i<(strlen(argv[0])+5); i++)
        {
            argv[0][i]='a';
        }
    }

    info=fopen(f, "a");
    if(info!=NULL)
    {
        writeData(info, 0, argc, argv);
        fclose(info);
    }
    else
    {
        printf("**FATAL ERROR\n");
        return 1;
    }


    /////////////////////////////////////////////////////////////////////////////////////////////////////////

    //ENCRYPTION

    /*
    ALGORITHM:

    Start word count at 1 and letter count at 0.
    Loop through every letter in the text

        Set the target letter to the first letter in the text.

        If the target letter is a space, print instructions on how to make a space in Coordinate Code and skip all proceeding steps.
        Otherwise:
            Loop through the key:
                Whenever a new letter is found, increase lettercount by 1.
                Whenever a new word is found (that is, if the current letter is a letter and the preceding letter is a space), increase word count and set word count to 0.
                If the current letter in the key matches the current key letter, print lettercount and wordcount.

        Set current key letter to the next letter in the text.
        Reset word and letter counts.
    */

    printf("\n");

    //determines if a letter has been found in the text. Will be set to true if the target letter is found
    bool targetLetterPresent=false;

    //start word count at 1, letter count at 0, declare target letter
    unsigned int word=1;
    unsigned int letter=0;

    //loop through every letter in the text:
    for(int t=0; t<=strlen(plaintext)-1; t++)
    {
        //Target letter is the current letter in the text, which is plaintext[t]
        printf("\n");

        //print header
        printf("%c: ", plaintext[t]);


        //If the target letter is a space, print instructions on how to make a space in Coordinate Code.
        //I believe the way spaces are written in Coordinate Code makes the code as a whole a lot easier to crack, even without knowing the key...
        if(plaintext[t]==' ')
        {
            printf("*To make a space, make either the x or y value equal 0.");
            targetLetterPresent=true;
        }
        //if not alphanumeric, print message
        else if(isalnum(plaintext[t])==false)
        {
            printf("**All non-alphanumeric characters count as spaces.");
            targetLetterPresent=true;
        }
        //If the target letter is a letter...
        else
        {
            //Bug fix. Decrease word count if the first letter of the key is a space
            if(isspace(key[0]))
            {
                word--;
            }

            //Loop through the key:
            for(int k=0; k<=strlen(key)-1; k++)
            {
                //Whenever a new letter is found, increase lettercount by 1.
                if(isalnum(key[k]))
                {
                    letter++;
                }

                //Whenever a new word is found (that is, if the current index is a letter and the preceding index is a space), increase word count by 1 and set word count to 0.
                if(key[k]!=' ' && key[k-1]==' ')
                {
                    word++;
                    letter=1;
                }

                //If the current letter in the key matches the current key letter, print lettercount and wordcount
                if(key[k] == plaintext[t])
                {
                    //If newline print is selected, print a new line
                    if(argc>1)
                    {
                        printf("\n");
                    }
                    //lettercount and wordcount
                    printf("(%i, %i) ", word, letter);
                    //note that a letter has been found
                    targetLetterPresent=true;
                }


            }


        }

        //if the target letter has not been found, print info
        if(targetLetterPresent==false)
        {
            printf("<Does not exist>");
        }

        printf("\n");

        //reset variables for next letter
        word=1;
        letter=0;
        targetLetterPresent=false;

        //Set current key letter to the next letter in the text. Already done in the outside for-loop "t=0; t<strlen(input)...".
    }

    //CONGRATUMALATIONS! YOU DID IT!
    printf("\n\n");

    return 0;
}

File: ccEncryptionAssist.py
#Python Coordinate Code Encryption Assist v1.0.1

#import all the stuff
import sys
import numpy
from cs50 import get_string
import datetime
import daniilkvyat
#include <stdio.h>
#include <cs50.h>
#include <daniilkvyat.h>


#NOTE: range(x,y) contains all numbers n such that: x <= n < y. DOES NOT INCLUDE Y!

#len but for numbers
def numberlen(number):
    output=0;

    if number<0:
        while True:
            number=number/10;
            output=output+1;

            if number>-1:
                return output+1;
    else:
        while True:
            number=number/10;
            output=output+1;

            if number<1:
                return output;



#lowercases everything and removes punctuation
def removePunctuation(inputText):
    #lowercase the input text
    inputText=inputText.lower()

    #start with empty output text
    outputText=""

    #src: https://appdividend.com/2020/11/02/python-string-append/#:~:text=To%20append%20a%20string%20in,to%20get%20the%20final%20string.

    #loop through input text
    for i in range(0,len(inputText)):

        #if input is alphanumeric, add the current letter in the input to the output string
        if inputText[i].isalnum():
            outputText = outputText + inputText[i]
        #if not and current character is aprostrophe, add a character that no one uses
        elif ord(inputText[i])==39 or ord(inputText[i])==96:
            outputText = outputText + "^"
        #otherwise, add a space to the string
        else:
            outputText = outputText + " "


    return outputText


#login function
def login(enteredPassword):

    #passwords
    passwords=numpy.array(["0", "12345"])

    passkeys=numpy.empty(len(passwords), dtype=numpy.bool_)

    for s in range(0, len(passkeys)):
        passkeys[s]=True


    for p in range(0, len(passwords)):
        if not enteredPassword==passwords[p]:
            passkeys[p]=False


    correctFakes=len(passwords)-1

    for c in range(0, len(passkeys)-1):
        if not passkeys[c]==True:
            correctFakes=correctFakes-1;


    if correctFakes<=0 and not passkeys[len(passwords)-1]==True:
        return 1

    elif correctFakes>0 and not passkeys[len(passwords)-1]==True:
        return -1

    return 0


#####################################################################################################

#I can name the main function whatever I want.
def runEncAssist():


    #FILES: https://stackabuse.com/writing-to-a-file-with-pythons-print-function/
    terminal=sys.stdout
    extFile=".data"



    #SECURITY

    #Determine login status from the user's entered password (get_string asks user to enter a password)
    password=get_string("Enter password: ")
    loginStatus=login(password)

    #Stop program if the user entered an incorrect password
    if loginStatus>0:
        try:
            f=open(extFile, "a");
            writeData(f, 2, sys.argv);
            f.close()
        except:
            print("**FATAL ERROR")
            return 1
        print("Incorrect password.")
        return 2

    #if not, accept the user
    print("Access granted.", end=' ')


    #reminder of newline print option
    if len(sys.argv)>1:
        print("Printing each ordered pair on a new line.")
    else:
        print("Note: enter 'python .ccEncryptionAssist.py n' to print each ordered pair on a new line")

    print()

    ################################################

    #INPUT

    #ask user for key and plaintext
    while True:
        key=get_string("Enter key: ");
        if len(key)>0:
            break

    print()

    while True:
        plaintext=get_string("Enter plaintext: ");
        if len(plaintext)>0:
            break

    #format key and plaintext
    key=removePunctuation(key)
    plaintext=plaintext.lower()

    ##################################################

    if not loginStatus==0:
        try:
            f=open(extFile, "a");
            writeData(f, 83, sys.argv);
            f.close()
        except:
            print("**FATAL ERROR")
            return 1
    else:
        try:
            f=open(extFile, "a");
            writeData(f, 0, sys.argv);
            f.close()
        except:
            print("**FATAL ERROR")
            return 1


    #DECODING
    print()

    word=1
    letter=0

    preKeyLetter='*'

    targetLetterPresent=False

    for plaintextLetter in plaintext:
        if not loginStatus==-1000:
            while True:
                word=word+1

        #print header
        print()
        print(plaintextLetter, end=': ')

        #print info if current letter in the plaintext is not alphanumeric
        if plaintextLetter==' ':
            print("To make a space, set either the X or Y coordinate to 0.", end='')
            targetLetterPresent=True

        elif plaintextLetter.isalnum()==False:
            print("**All non-alphanumeric characters count as spaces.", end='')
            targetLetterPresent=True

        #if is alphanumeric, start
        else:

            for keyLetter in key:

                #increase letters if next character is a letter
                if keyLetter.isalnum():
                    letter=letter+1

                #increase words and reset letters if new word is found in the key
                if preKeyLetter==' ' and not keyLetter==' ':
                    word=word+1
                    letter=1

                #print ordered pair when match is found
                if keyLetter==plaintextLetter:

                    #print newline if newline option is selected
                    if len(sys.argv)>1:
                        print()

                    print("(" , end='')
                    print(word , end='')
                    print(", " , end='')
                    print(letter , end='')
                    print(") " , end='')

                    #confirm the find
                    targetLetterPresent=True

                #set preKeyLetter to the previous letter compared
                preKeyLetter=keyLetter




        #notify if target letter has not been found
        if targetLetterPresent==False:
            print("<Does not exist>", end='')

        #reset counts
        word=1
        letter=0
        preKeyLetter='*'
        targetLetterPresent=False



        #newline
        print()



def writeData(file, retCode, argv):
    file.write(str(retCode));
    for i in range(0, 5-numberlen(retCode)):
        file.write(" ");

    if len(argv)<=999:
        file.write(str(len(argv)));
        for i in range(0, 5-numberlen(len(argv))):
            file.write(" ");
    else:
        file.write("999  ");

    if len(argv[0])<=30:
        file.write(argv[0]);
        for i in range(0, 32-len(argv[0])):
            file.write(" ");
    else:
        for i in range(0, 30):
            file.write(argv[0][i]);
        file.write("  ");

    file.write(str(datetime.datetime.now()));

    file.write("\n");

#call main function
if __name__ == "__main__":
    runEncAssist()